### ISSUE TITLE
classpath should be in the same order as maven

### ISSUE THREAD

**Issue by maschmid** [fabric8io/java-main#11]
Currently the run script generates classpath via "ls $DEPLOY_DIR/dependency/*.jar", which will generate classpath in a different order than maven would do with "mvn exec:java".

**Comment by chirino** [comment]
I think these examples need to reworked to be packaged using the hawt-app mvn plugin found at https://github.com/jboss-fuse/hawt-app.  This way you do get the same classpath order as mvn exec:java.

**Comment by jimmidyson** [comment]
I believe @rhuss has already sorted this for Docker Maven Plugin.

**Comment by chirino** [comment]
Still would be good to use the hawt-app stuff so that you can get working distro of the app inside or outside of OS3.

**Comment by jimmidyson** [comment]
d-m-p is unrelated to OS3, just produces Docker images which is definitely the primary use case for this. Producing a tarball is also in the pipeline for d-m-p I believe. Again, I defer to @rhuss on this for further clarification.

**Comment by jstrachan** [comment]
up to now fabric8 quickstarts has used the jube plugin to generate a tarball (zip actually which is the JBoss distro standard) for stand alone non-docker usage; then docker maven plugin for docker images and fabric8 for the k8s resources. The examples all still have jube support for stand alone distros.

If anything it'd be better if we combined maven plugins together so there's less things to configure on a per project basis as the main things a java project needs to specify (its main class and dependencies) tend to be the same. This would also help simplify the forge commands to fabric8-enable projects etc (which right now only support fabric8 plugin and docker plugin).

So I'd like to see fabric8 + docker maven plugins merge long term; along with combining the tar/zip generation into the docker plugin

**Comment by rhuss** [comment]
@maschmid the [run.sh](https://github.com/fabric8io/base-images/blob/master/blocks/run-java-sh/run-java.sh#L17-L25) the in the new base image honor the existance of a `classpath` file which for the STI base image is created during the [assembly](https://github.com/fabric8io/base-images/blob/master/sti/templates/sti/assemble#L61-L64). This file can also easily be created with the dependency plugin in a regular build which then can be included into a d-m-p assembly. I will add an example soon with plain d-m-p and maven-dependency usage.

@chirino agree that the plain-tar-delivery model is useful (as the jube plugin provides). I'm not sure whether you really need an own plugin for a tar-with-run.sh creation. Also, I'm still not convinced that the run script belongs to the maven plugin. Nevertheless, what do you think about putting the run script into a separate maven project (e.g. fabric8io/java-run-sh) which can be referenced via a maven dep by your plugin and via github reference by the base images ? I will code a prototype till tomorrow so that we can discuss that there. 

@jstrachan I agree that we shouldn't fragment the maven plugin space too much and I also think that we could put the tar-file creation into d-m-p maybe with an extra `<jube>` section for the build config.

**Comment by chirino** [comment]
So generating tar balls is a separate concern from docker images.  That' the main thing I'm trying to keep distinct.  The fabric8 mvn plugin seems mostly associated with OS3, which application tar balls are kinda different from too.  So I still think it make sense to keep this as a separate plugin.

**Comment by chirino** [comment]
@rhuss so one day Fuse might want to support this micro-service model of development outside of OS3, and in that case users probably will not be using fabric, or docker or any of this stuff.  But they will still want to build an app .tag.gz with start scripts.  That's why it makes sense to keep it separate and include the start scripts as part of the distro generation.

**Comment by jstrachan** [comment]
@chirino from the application developer there's only so many times you wanna configure a mvn plugin to say what the main class is and what includes/excludes/custom assembly XML you wanna use to tweak what jars get used. There's huge overlap between docker images / k8s metadata / tarballs in terms of user configuration. I don't see why 1 mvn plugin can't do all 3 in a single canonical way of defining what artifacts get included; what main is used, what env vars to define etc

**Comment by jimmidyson** [comment]
@jstrachan :+1: 

@chirino "All" the d-m-p does is create a tarball & layer it into a docker image. That tarball could be used as is AFAIK, without the Docker build part...

**Comment by chirino** [comment]
@jstrachan I could be swayed to put it all in one plugin to reduce config settings.  Seems a little smelly, but what the hell.

**Comment by jstrachan** [comment]
if there was a way in maven to reuse configuration across mvn plugins we could keep them separate (its what we did with jube + docker plugins even though they are pretty much identical apart from whether a file is added to the project or a tar is posted). 

Until then using a single maven plugin is simpler for users and just having different goals to differentiate the bits folks wanna do. 

e.g. so anyone could create a tarball on any fabric8 example without touching a line of pom.xml.

**Comment by rhuss** [comment]
I agree with @chirino that putting sth non-docker related into a docker-maven-plugin sounds weird and feels wrong. 

IMO there are two solutions: Narrowing the scope by providing multiple plugins with a lot of overlap /wrt the configuration (and the danger that they have similar syntax but still slightly different) or widening the scope. E.g. putting it into something somewhat 'docker agnostic' (at least /wrt the naming :) like a fabric8 plugin.

But maybe there is even another road that we can go in order to avoid duplication (which is my main intention) ? What I really would like is to share common functionality like the run scripts.
